File: src/ObjectManager/Object.Tes/FilePacks/Records/345-REGN.Region.cs
using OA.Core;
using System.Collections.Generic;
using UnityEngine;

namespace OA.Tes.FilePacks.Records
{
    public class REGNRecord : Record, IHaveEDID
    {
        // TESX
        public class RDATField
        {
            public enum REGNType : byte
            {
                Objects = 2, Weather, Map, Landscape, Grass, Sound
            }

            public uint Type;
            public REGNType Flags;
            public byte Priority;
            // groups
            public RDOTField[] RDOTs; // Objects
            public STRVField RDMP; // MapName
            public RDGSField[] RDGSs; // Grasses
            public UI32Field RDMD; // Music Type
            public RDSDField[] RDSDs; // Sounds
            public RDWTField[] RDWTs; // Weather Types

            public RDATField() { }
            public RDATField(UnityBinaryReader r, int dataSize)
            {
                Type = r.ReadLEUInt32();
                Flags = (REGNType)r.ReadByte();
                Priority = r.ReadByte();
                r.SkipBytes(2); // Unused
            }
        }

        public struct RDOTField
        {
            public override string ToString() => $"{Object}";
            public FormId<Record> Object;
            public ushort ParentIdx;
            public float Density;
            public byte Clustering;
            public byte MinSlope; // (degrees)
            public byte MaxSlope; // (degrees)
            public byte Flags;
            public ushort RadiusWrtParent;
            public ushort Radius;
            public float MinHeight;
            public float MaxHeight;
            public float Sink;
            public float SinkVariance;
            public float SizeVariance;
            public Vector3Int AngleVariance;
            public ColorRef VertexShading; // RGB + Shading radius (0 - 200) %

            public RDOTField(UnityBinaryReader r, int dataSize)
            {
                Object = new FormId<Record>(r.ReadLEUInt32());
                ParentIdx = r.ReadLEUInt16();
                r.SkipBytes(2); // Unused
                Density = r.ReadLESingle();
                Clustering = r.ReadByte();
                MinSlope = r.ReadByte();
                MaxSlope = r.ReadByte();
                Flags = r.ReadByte();
                RadiusWrtParent = r.ReadLEUInt16();
                Radius = r.ReadLEUInt16();
                MinHeight = r.ReadLESingle();
                MaxHeight = r.ReadLESingle();
                Sink = r.ReadLESingle();
                SinkVariance = r.ReadLESingle();
                SizeVariance = r.ReadLESingle();
                AngleVariance = new Vector3Int(r.ReadLEUInt16(), r.ReadLEUInt16(), r.ReadLEUInt16());
                r.SkipBytes(2); // Unused
                VertexShading = new ColorRef(r);
            }
        }

        public struct RDGSField
        {
            public override string ToString() => $"{Grass}";
            public FormId<GRASRecord> Grass;

            public RDGSField(UnityBinaryReader r, int dataSize)
            {
                Grass = new FormId<GRASRecord>(r.ReadLEUInt32());
                r.SkipBytes(4); // Unused
            }
        }

        public struct RDSDField
        {
            public override string ToString() => $"{Sound}";
            public FormId<SOUNRecord> Sound;
            public uint Flags;
            public uint Chance;

            public RDSDField(UnityBinaryReader r, int dataSize, GameFormatId format)
            {
                if (format == GameFormatId.TES3)
                {
                    Sound = new FormId<SOUNRecord>(r.ReadASCIIString(32, ASCIIFormat.ZeroPadded));
                    Flags = 0;
                    Chance = r.ReadByte();
                    return;
                }
                Sound = new FormId<SOUNRecord>(r.ReadLEUInt32());
                Flags = r.ReadLEUInt32();
                Chance = r.ReadLEUInt32(); //: float with TES5
            }
        }

        public struct RDWTField
        {
            public override string ToString() => $"{Weather}";
            public static byte SizeOf(GameFormatId format) => format == GameFormatId.TES4 ? (byte)8 : (byte)12;
            public FormId<WTHRRecord> Weather;
            public uint Chance;
            public FormId<GLOBRecord> Global;

            public RDWTField(UnityBinaryReader r, int dataSize, GameFormatId format)
            {
                Weather = new FormId<WTHRRecord>(r.ReadLEUInt32());
                Chance = r.ReadLEUInt32();
                Global = format == GameFormatId.TES5 ? new FormId<GLOBRecord>(r.ReadLEUInt32()) : new FormId<GLOBRecord>();
            }
        }

        // TES3
        public struct WEATField
        {
            public byte Clear;
            public byte Cloudy;
            public byte Foggy;
            public byte Overcast;
            public byte Rain;
            public byte Thunder;
            public byte Ash;
            public byte Blight;

            public WEATField(UnityBinaryReader r, int dataSize)
            {
                Clear = r.ReadByte();
                Cloudy = r.ReadByte();
                Foggy = r.ReadByte();
                Overcast = r.ReadByte();
                Rain = r.ReadByte();
                Thunder = r.ReadByte();
                Ash = r.ReadByte();
                Blight = r.ReadByte();
                // v1.3 ESM files add 2 bytes to WEAT subrecords.
                if (dataSize == 10)
                    r.SkipBytes(2);
            }
        }

        // TES4
        public class RPLIField
        {
            public uint EdgeFalloff; // (World Units)
            public Vector2[] Points; // Region Point List Data

            public RPLIField(UnityBinaryReader r, int dataSize)
            {
                EdgeFalloff = r.ReadLEUInt32();
            }

            public void RPLDField(UnityBinaryReader r, int dataSize)
            {
                Points = new Vector2[dataSize >> 3];
                for (var i = 0; i < Points.Length; i++) Points[i] = new Vector2(r.ReadLESingle(), r.ReadLESingle());
            }
        }

        public override string ToString() => $"REGN: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public STRVField ICON; // Icon / Sleep creature
        public FMIDField<WRLDRecord> WNAM; // Worldspace - Region name
        public CREFField RCLR; // Map Color (COLORREF)
        public List<RDATField> RDATs = new List<RDATField>(); // Region Data Entries / TES3: Sound Record (order determines the sound priority)
        // TES3
        public WEATField? WEAT; // Weather Data
        // TES4
        public List<RPLIField> RPLIs = new List<RPLIField>(); // Region Areas

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "WNAM":
                case "FNAM": WNAM = new FMIDField<WRLDRecord>(r, dataSize); return true;
                case "WEAT": WEAT = new WEATField(r, dataSize); return true; //: TES3
                case "ICON":
                case "BNAM": ICON = new STRVField(r, dataSize); return true;
                case "RCLR":
                case "CNAM": RCLR = new CREFField(r, dataSize); return true;
                case "SNAM": RDATs.Add(new RDATField { RDSDs = new[] { new RDSDField(r, dataSize, format) } }); return true;
                case "RPLI": RPLIs.Add(new RPLIField(r, dataSize)); return true;
                case "RPLD": ArrayUtils.Last(RPLIs).RPLDField(r, dataSize); return true;
                case "RDAT": RDATs.Add(new RDATField(r, dataSize)); return true;
                case "RDOT":
                    var rdot = ArrayUtils.Last(RDATs).RDOTs = new RDOTField[dataSize / 52];
                    for (var i = 0; i < rdot.Length; i++) rdot[i] = new RDOTField(r, dataSize); return true;
                case "RDMP": ArrayUtils.Last(RDATs).RDMP = new STRVField(r, dataSize); return true;
                case "RDGS":
                    var rdgs = ArrayUtils.Last(RDATs).RDGSs = new RDGSField[dataSize / 8];
                    for (var i = 0; i < rdgs.Length; i++) rdgs[i] = new RDGSField(r, dataSize); return true;
                case "RDMD": ArrayUtils.Last(RDATs).RDMD = new UI32Field(r, dataSize); return true;
                case "RDSD":
                    var rdsd = ArrayUtils.Last(RDATs).RDSDs = new RDSDField[dataSize / 12];
                    for (var i = 0; i < rdsd.Length; i++) rdsd[i] = new RDSDField(r, dataSize, format); return true;
                case "RDWT":
                    var rdwt = ArrayUtils.Last(RDATs).RDWTs = new RDWTField[dataSize / RDWTField.SizeOf(format)];
                    for (var i = 0; i < rdwt.Length; i++) rdwt[i] = new RDWTField(r, dataSize, format); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-REGN.Region.swift
//
//  REGNRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

import CoreGraphics
import simd

public class REGNRecord: Record, IHaveEDID {
    // TESX
    public class RDATField {
        public enum REGNType: UInt8 {
            case objects = 2, weather, map, landscape, grass, sound
        }

        public var type: UInt32 = 0
        public var flags: REGNType = .objects
        public var priority: UInt8 = 0
        // groups
        public var RDOTs: [RDOTField]! // Objects
        public var RDMP: STRVField! // MapName
        public var RDGSs: [RDGSField]! // Grasses
        public var RDMD: UI32Field! // Music Type
        public var RDSDs: [RDSDField]! // Sounds
        public var RDWTs: [RDWTField]! // Weather Types

        init(RDSDs: [RDSDField]) {
            self.RDSDs = RDSDs
        }
        init(_ r: BinaryReader, _ dataSize: Int) {
            type = r.readLEUInt32()
            flags = REGNType(rawValue: r.readByte())!
            priority = r.readByte()
            r.skipBytes(2) // Unused
        }
    }

    public struct RDOTField: CustomStringConvertible {
        public var description: String { return "\(object)" }
        public let object: FormId<Record>
        public let parentIdx: UInt16
        public let density: Float
        public let clustering: UInt8
        public let minSlope: UInt8 // (degrees)
        public let maxSlope: UInt8 // (degrees)
        public let flags: UInt8
        public let radiusWrtParent: UInt16
        public let radius: UInt16
        public let minHeight: Float
        public let maxHeight: Float
        public let sink: Float
        public let sinkVariance: Float
        public let sizeVariance: Float
        public let angleVariance: int3
        public let vertexShading: ColorRef4 // RGB + Shading radius (0 - 200) %

        init(_ r: BinaryReader, _ dataSize: Int) {
            object = FormId<Record>(r.readLEUInt32())
            parentIdx = r.readLEUInt16()
            r.skipBytes(2) // Unused
            density = r.readLESingle()
            clustering = r.readByte()
            minSlope = r.readByte()
            maxSlope = r.readByte()
            flags = r.readByte()
            radiusWrtParent = r.readLEUInt16()
            radius = r.readLEUInt16()
            minHeight = r.readLESingle()
            maxHeight = r.readLESingle()
            sink = r.readLESingle()
            sinkVariance = r.readLESingle()
            sizeVariance = r.readLESingle()
            angleVariance = int3(Int32(r.readLEUInt16()), Int32(r.readLEUInt16()), Int32(r.readLEUInt16()))
            r.skipBytes(2) // Unused
            vertexShading = r.readT(dataSize)
        }
    }

    public struct RDGSField: CustomStringConvertible {
        public var description: String { return "\(grass)" }
        public let grass: FormId<GRASRecord>

        init(_ r: BinaryReader, _ dataSize: Int) {
            grass = FormId<GRASRecord>(r.readLEUInt32())
            r.skipBytes(4) // Unused
        }
    }

    public struct RDSDField: CustomStringConvertible {
        public var description: String { return "\(sound)" }
        public let sound: FormId<SOUNRecord> 
        public let flags: UInt32
        public let chance: UInt32

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format != .TES3 else {
                sound = FormId<SOUNRecord>(r.readASCIIString(32, format: .zeroPadded))
                flags = 0
                chance = UInt32(r.readByte())
                return
            }
            sound = FormId<SOUNRecord>(r.readLEUInt32())
            flags = r.readLEUInt32()
            chance = r.readLEUInt32() //: float with TES5
        }
    }

    public struct RDWTField: CustomStringConvertible {
        public var description: String { return "\(weather)" }
        public static func sizeOf(for format: GameFormatId) -> Int { return format == .TES4 ? 8 : 12 }
        public let weather: FormId<WTHRRecord>
        public let chance: UInt32
        public let global: FormId<GLOBRecord>

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            weather = FormId<WTHRRecord>(r.readLEUInt32())
            chance = r.readLEUInt32()
            global = format == .TES5 ? FormId<GLOBRecord>(r.readLEUInt32()) : FormId<GLOBRecord>(0)
        }
    }

    // TES3
    public typealias WEATField = (
        clear: UInt8,
        cloudy: UInt8,
        foggy: UInt8,
        overcast: UInt8,
        rain: UInt8,
        thunder: UInt8,
        ash: UInt8,
        blight: UInt8
    )
    
    // TES4
    public class RPLIField {
        public let edgeFalloff: UInt32 // (World Units)
        public var points: [float2]! // Region Point List Data

        init(_ r: BinaryReader, _ dataSize: Int) {
            edgeFalloff = r.readLEUInt32()
        }

        func RPLDField(_ r: BinaryReader, _ dataSize: Int) {
            points = r.readTArray(dataSize, count: dataSize >> 3)
        }
    }

    public override var description: String { return "REGN: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var ICON: STRVField! // Icon / Sleep creature
    public var WNAM: FMIDField<WRLDRecord>! // Worldspace - Region name
    public var RCLR: CREFField! // Map Color (COLORREF)
    public var RDATs = [RDATField]() // Region Data Entries / TES3: Sound Record (order determines the sound priority)
    // TES3
    public var WEAT: WEATField? = nil // Weather Data
    // TES4
    public var RPLIs = [RPLIField]() // Region Areas

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "WNAM",
             "FNAM": WNAM = FMIDField<WRLDRecord>(r, dataSize)
        case "WEAT": WEAT = r.readT(dataSize)
        case "ICON",
             "BNAM": ICON = r.readSTRV(dataSize)
        case "RCLR",
             "CNAM": RCLR = r.readT(dataSize)
        case "SNAM": RDATs.append(RDATField(RDSDs: [RDSDField(r, dataSize, format)]))
        case "RPLI": RPLIs.append(RPLIField(r, dataSize))
        case "RPLD": RPLIs.last!.RPLDField(r, dataSize)
        case "RDAT": RDATs.append(RDATField(r, dataSize))
        case "RDOT":
            var rdot = [RDOTField](); RDATs.last!.RDOTs = rdot; let capacity = dataSize / 52; rdot.reserveCapacity(capacity)
            for _ in 0..<capacity { rdot.append(RDOTField(r, dataSize)) }
        case "RDMP": RDATs.last!.RDMP = r.readSTRV(dataSize)
        case "RDGS":
            var rdgs = [RDGSField](); RDATs.last!.RDGSs = rdgs; let capacity = dataSize / 8; rdgs.reserveCapacity(capacity)
            for _ in 0..<capacity { rdgs.append(RDGSField(r, dataSize)) }
        case "RDMD": RDATs.last!.RDMD = r.readT(dataSize)
        case "RDSD":
            var rdsd = [RDSDField](); RDATs.last!.RDSDs = rdsd; let capacity = dataSize / 15; rdsd.reserveCapacity(capacity)
            for _ in 0..<capacity { rdsd.append(RDSDField(r, dataSize, format)) }
        case "RDWT":
            var rdwt = [RDWTField](); RDATs.last!.RDWTs = rdwt; let capacity = dataSize / RDWTField.sizeOf(for: format); rdwt.reserveCapacity(capacity)
            for _ in 0..<capacity { rdwt.append(RDWTField(r, dataSize, format)) }
        default: return false
        }
        return true
    }
}
